src/bin/sol-fbp-runner/Widget.js

```diff
@@ -0,0 +1,222 @@
+		this.totalPorts = 0;
+
+		//TODO: we should change the id key to name
+		this.name = "widget" + payload.id;
+
+		//title and subtitle of the widget
+		this.title = $("<span/>").addClass("widget-title").append(this.widgetName);
+		this.subtitle = $("<span/>").addClass("widget-subtitle").append(this.type);
+
+		//info button
+		var info = $("<img/>").attr("src","images/ico_info.png").addClass("info");
+		info.click(function(event) {
+			window.modal.show(instance.raw, instance.type);
+		});
+
+		//creating DOM elements
+		this.displayValue = $("<span/>").addClass("display-value");
+		this.el = $("<div/>").addClass("widget").attr("id",this.uid).appendTo("#container");
+		$("<div/>").addClass("widget-header").appendTo(this.el).append(this.title).append(this.subtitle).append(info);
+		$("<div/>").addClass("widget-display").appendTo(this.el).addClass("widget-data").append(this.displayValue);
+		this.widgetPorts = $("<div/>").addClass("widget-ports").appendTo(this.el);
+
+		//putting the ports in the correct container
+		this.lports = $("<div/>").addClass("port-container").appendTo("#"+this.uid + " .widget-ports");
+		this.rports = $("<div/>").addClass("port-container").appendTo("#"+this.uid + " .widget-ports");
+
+		//reading all in ports
+		var port;
+		payload.ports_in.forEach(function(element, index) {
+			port = new Port(
+			{
+				name:element["name"],
+				value:undefined,
+				type:"in",
+				port_index:element["base_port_idx"],
+				data_type:element["data_type"],
+				required:element["required"],
+				widget:instance,
+				key:instance.key + "_" + Widget.prototype.portsCount
+			});
+
+			instance.portsIn.push(port);
+			instance.totalPorts++;
+			port.el.on("port-select", instance.onPortSelect);
+			Widget.prototype.portsCount ++;
+		});
+
+		//reading all out ports
+		payload.ports_out.forEach(function(element, index) {
+			port = new Port(
+			{
+				name:element["name"],
+				value:undefined,
+				type:"out",
+				port_index:element["base_port_idx"],
+				data_type:element["data_type"],
+				required:element["required"],
+				widget:instance,
+				key:instance.key + "_" + Widget.prototype.portsCount
+			});
+			instance.portsOut.push(port);
+			instance.totalPorts++;
+			port.el.on("port-select", instance.onPortSelect);
+			Widget.prototype.portsCount ++;
+		});
+
+		//Auto Adjusting the size of the Widget Container
+		var h = Math.max(this.portsIn.length, this.portsOut.length) * 26;
+		this.widgetPorts.css("height",h);
+
+	}
+
+	Widget.prototype.onPortSelect = function(event) {
+		event.widget.selectPort(event.port, event.value);
+	};
+
+	Widget.prototype.getUID = function() {
+		return this.uid;
+	};
+
+	//UPDATES THE VALUES OF ALL PORTS
+	Widget.prototype.update = function(type, data) {
+
+		//getting the child ports of the widget
+		var port = (type === "deliver") ? this.portsIn[data.port_idx] : this.portsOut[data.port_idx];
+
+		//if the ports have no pair, skip
+		if(port.connections.length < 0) return;
```

Comment:
for Inspector, it may be useful to know the packet being delivered even if there are no connections. I'd not do this early return